lms/djangoapps/shoppingcart/models.py

```diff
@@ -1606,3 +1609,355 @@ def analytics_data(self):
+
+    # any 'account identified' (aka Merchant Account) associated with this transaction
+    # note, this is an optional field
+    account_id = models.CharField(max_length=32, null=True)
+
+    # The timestamp associated with when the payment processing
+    # occured
+    processed_at = models.DateTimeField(db_index=True)
+
+    # the Order model that this processing is associated with
+    order = models.ForeignKey(Order, db_index=True)
+
+    currency = models.CharField(max_length=16)
+    amount = models.DecimalField(decimal_places=2, max_digits=30)
+
+    # transaction_type in Payment Processing report is
+    # whether it is a purchase or a refund: 'purchase' or 'refund'
+    TRANSACTION_TYPE_CHOICES = (
+        (TRANSACTION_TYPE_PURCHASE, ugettext_noop('Purchase')),
+        (TRANSACTION_TYPE_REFUND, ugettext_noop('Refund'))
+    )
+    transaction_type = models.CharField(max_length=6, db_index=True, choices=TRANSACTION_TYPE_CHOICES)
+
+    def __eq__(self, other):
+        """
+        Equality operator
+        """
+
+        return (
+            unicode(self.remote_transaction_id) == unicode(other.remote_transaction_id) and
+            unicode(self.account_id) == unicode(other.account_id) and
+            self.processed_at == other.processed_at and
+            self.order.id == other.order.id and  # pylint: disable=no-member
+            unicode(self.currency) == unicode(other.currency) and
+            self.amount == other.amount and
+            self.transaction_type == other.transaction_type
+        )
+
+    @classmethod
+    def create(cls, remote_transaction_id, account_id, processed_at, order_id, currency, amount, transaction_type):
+        """
+        This classmethod will create a new entry in the PaymentTransaction table
+        It will use a get_or_create() at the ORM level.
+        We enforce that a Transaction is immutable and we therefore will
+        not support update cases
+        """
+
+        try:
+            order = Order.objects.get(id=order_id)
+        except Order.DoesNotExist:
+            msg = (
+                'Transaction {transaction_id} refers to order_id {order_id} '
+                'but it cannot be found!'
+            ).format(
+                transaction_id=remote_transaction_id,
+                order_id=order_id,
+            )
+            raise OrderDoesNotExistException(msg)
+
+        cc_transaction = PaymentTransaction(
+            remote_transaction_id=remote_transaction_id,
+            account_id=account_id,
+            processed_at=processed_at,
+            order=order,
+            currency=currency,
+            amount=amount,
+            transaction_type=transaction_type
+        )
+
+        # Do a quick check to see if it exists
+        exists = PaymentTransaction.objects.filter(remote_transaction_id=remote_transaction_id).exists()
+
+        if exists:
+            # see if an already existing row is *exactly* the same
```

Comment:
Would it be possible to enforce this as a database constraint?  I'm worried about a potential race condition if a payment is saved after this transaction opens.